haskell-mode.el

```diff
@@ -501,6 +501,63 @@ executable found in PATH.")
     table)
   "Syntax table used in Haskell mode.")
 
+(defun haskell-find-type-end (&optional ignore)
+  "Find where does this type declaration end.
+
+Moves the point to the end of type declaration. It should be
+invoked with point just after one of type introducing keywords
+like ::, class, instance, data, newtype, type."
+  (interactive)
+  (let ((cont t)
+        (end (point))
+        (token nil)
+        ;; we are starting right after ::
+        (last-token-was-operator t)
+        (last-token-was-newline nil))
+    (while cont
+      (setq token (haskell-lexeme-looking-at-token 'newline))
+
+      (cond
+       ((null token)
+        (setq cont nil))
+       ((member token '(newline))
+        (setq last-token-was-newline (not last-token-was-operator))
+        (goto-char (match-end 0)))
+       ((and (member (match-string-no-properties 0)
+                     '("<-" "=" "<-" "←"  "," ";"
+                       ")" "]" "}" "|"
+                       "case" "class" "data" "default" "deriving" "do"
```

Comment:
And, btw, thanks for reading this. It wasn't easy to write so I'm grateful for all the help.